db/migrations/20220608090500_remove_duplicated_tenant_ids_org_ids.go

```diff
@@ -0,0 +1,255 @@
+package migrations
+
+import (
+	"fmt"
+
+	logging "github.com/RedHatInsights/sources-api-go/logger"
+	"github.com/RedHatInsights/sources-api-go/model"
+	"github.com/go-gormigrate/gormigrate/v2"
+	"gorm.io/gorm"
+)
+
+// duplicateTenantMark is the mark that will be left on the "external_tenant" and "org_id" columns for those tenants
+// that are duplicated. The idea is to prepare the tenants to be able to apply the unique constraint on the columns
+// right after.
+const duplicateTenantMark = "processed-duplicate-tenant-of-%s-%d"
+
+// RemoveDuplicatedTenantIdsOrgIds fetches all the duplicated tenants, makes sure that they are indeed duplicated, and
+// unifies all the subresources in a single tenant before deleting the duplicated one.
+func RemoveDuplicatedTenantIdsOrgIds() *gormigrate.Migration {
+	return &gormigrate.Migration{
+		ID: "20220608090500",
+		Migrate: func(db *gorm.DB) error {
+			logging.Log.Info(`Migration "remove duplicated tenants" started`)
+			defer logging.Log.Info(`Migration "remove duplicated tenants" ended`)
+
+			// Updatable tables contains all the table names which have the "tenant_id" column which must be updated.
+			updatableTables := []string{
+				"application_authentications",
+				"applications",
+				"authentications",
+				"endpoints",
+				"sources",
+				"source_rhc_connections",
+				"users",
+			}
+
+			// This query selects all the tenants which have duplicated "external_tenant" or "org_id" column values.
+			duplicatesSql := `
+				SELECT
+					"t"."id", "t"."name", "t"."description", "t"."external_tenant", "t"."org_id", "t"."created_at", "t"."updated_at"
+				FROM
+					"tenants" AS "t"
+				WHERE
+					"t"."external_tenant" IN (
+						SELECT
+							"texten"."external_tenant"
+						FROM
+							"tenants" AS "texten"
+						GROUP BY
+							"texten"."external_tenant"
+						HAVING
+							count("texten"."external_tenant") > 1
+						AND
+							"texten"."external_tenant" != ''
```

Comment:
do we also need to cover when `external_tenant` is not `null` ?

Reply:
Actually I think we can remove the `"texten"."external_tenant" != ''`, since the previous migration takes care of doing the `'' -> NULL` conversion on the database.

As to your question: no, we don't need that condition. The `AND "external_tenant" != ''` was there because the empty string counted as a "duplicate" as well. So, if we had the following:



id | external_tenant | org_id
-- | -- | --
1 | 12345 | 9876543210
2 |   | abcde
3 |   | deeehg

It would pick up rows with id `2` and `3` as if they were a duplicate, when in reality the only duplicate there is the empty string on the `external_tenant` column. With `NULL`s that doesn't happen because `NULL == NULL -> false`. You can test this yourself:

```sql
INSERT INTO
	"tenants"("external_tenant", "org_id", "created_at", "updated_at")
VALUES
	(NULL, 'abcde', now(), now()),
	(NULL, 'edcba', now(), now()),
	('', 'hello', now(), now()),
	('', 'notsame', now(), now())
```

And then run the query without the `AND` condition:


```sql
SELECT
	"t"."id", "t"."name", "t"."description", "t"."external_tenant", "t"."org_id", "t"."created_at", "t"."updated_at"
FROM
	"tenants" AS "t"
WHERE
	"t"."external_tenant" IN (
		SELECT
			"texten"."external_tenant"
		FROM
			"tenants" AS "texten"
		GROUP BY
			"texten"."external_tenant"
		HAVING
			count("texten"."external_tenant") > 1
	)
OR
	"t"."org_id" IN (
		SELECT
			"torgid"."org_id"
		FROM
			"tenants" AS "torgid"
		GROUP BY
			"torgid"."org_id"
		HAVING
			count("torgid"."org_id") > 1
	)
```

You'll see that the query picks up the two rows with the empty string, when it shouldn't. However, as I said, in the previous migration I take care of the empty strings on both columns, and also I ran a query on AppInterface on the stage and prod environments to check if we have empty strings on either of the columns and we don't.

Reply:
thanks for detailed explanation, now I understand

Reply:
Here are the results of the queries that I ran on both clusters: https://issues.redhat.com/browse/RHCLOUD-19802 (restricted to the team for confidentiality reasons). The MR that I sent to AppInterface checked for `external_tenant = ''`, `"external_tenant" = '' AND "org_id" != ''` and `"external_tenant" = '' AND "org_id" = ''` and yielded zero results, so we should be safe.